ReactでToDo Listを作る

yuyakondo333 opened the pull request:
## 課題のリンク

* https://kirara-code.net/HappinessChain/tasks/59?chapterId=21

## やったこと

* ReactでToDo Listを作成


## 動作確認方法

### Create
- 画面中央の入力部分に文字を入力して追加ボタンを押下するとタスクが作成される
  - 空白、スペースだけで追加ボタンを押下しても何も起きない
### Read
- タスクは以下の構成要素になっている
  - 完了 / 未完了 のチェックボックス
  - タイトル
  - 編集ボタン
  - 削除ボタン
- タスク状況
  - タスクが追加されると、**全てのタスクのカウントが「1」増える**
  - チェックボックスで完了 / 未完了 を操作することで、**完了済み/ 未完了 のタスクの数が増減する**
  - タスク削除によって**タスク状況のカウントが減る**
### Update
- 任意のタスクの編集ボタンを押下すると編集モードに変わる
- タイトルを変更して更新ボタンを押下すると、タイトルが変更されている
- タイトルを変更するしない問わず、戻るボタンを押下すると編集モードから通常のモードに戻りタイトルは編集されない
- 編集モードでは他のタスクへのアクションはできない
### Delete
- 任意のタスクの削除ボタンを押下するとタスクが削除される
## その他
* 特になし

yuyakondo333 committed:
Todo app operation confirmed

yuyakondo333 committed:
split into components

yuyakondo333 committed:
Add placeholder

yuyakondo333 committed:
Preparation for deployment

yuyakondo333 committed:
Add predeploy

yuyakondo333 committed:
Modify the "deploy" command in package.json

yuyakondo333 committed:
Remove unused arguments

yuyakondo333 committed:
Implementing a confirmation dialog for deleting a task

yamaryo416 commented on src/App.jsx:
```diff
@@ -1,33 +1,100 @@
+    );
+  }
+
+  const completedTodos = (todos) => todos.filter(todo => todo.completed).length;
+  const incompletedTodos = (todos) => todos.filter(todo => !todo.completed).length;
```
こちらは関数ではなく、定数として定義しましょう。
```
const completedTodos = todos.filter(todo => todo.completed).length;
const incompletedTodos = todos.filter(todo => !todo.completed).length;
```

yuyakondo333 replied:
関数から定数に修正しました。

yamaryo416 commented on src/App.jsx:
```diff
@@ -1,33 +1,100 @@
+
+  const addTodoText = (event) => setInput(event.target.value);
+
+  const addTodo = () => {
+    if (input.trim()) {
+      setTodos([...todos, {
+        id: nextId.current++,
+        text: input,
+        completed: false,
+      }])
+      setInput('');
+    }
+  }
```
早期エラーリターンするのが良いですね。
```
const text = input.trim()
if (!text) return
setTodos([...todos, {
  id: nextId.current++,
  text,
  completed: false,
}])
setInput('');
```

yuyakondo333 replied:
early returnするコードに修正しました。

yamaryo416 commented on src/App.jsx:
```diff
@@ -1,33 +1,100 @@
-        </p>
+      <div>
+        <ul>
+        {todos.map((todo) => {
+          return (
+            <TodoItem
+              todo={todo}
+              editProps={{
```
mapで回す際は、keyを必ず与えましょう。
また、returnは省略できます。
```
{todos.map((todo) => (
  <TodoItem
    key= {todo.id}
```

yuyakondo333 replied:
keyを追記して return を省略しました。

yamaryo416 commented on src/App.jsx:
```diff
@@ -1,33 +1,100 @@
 
 function App() {
-  const [count, setCount] = useState(0)
+  const [todos, setTodos] = useState([]);
+  const [input, setInput] = useState('');
+  const [isEditing, setIsEditing] = useState(false);
+  const [currentTodo, setCurrentTodo] = useState(null);
+  const nextId = useRef(1);
```
`currentTodo`があれば、`isEditing`がなくても編集中かどうかわかるので、
`isEditing`は削除しても良いですね

yuyakondo333 replied:
currentTodo があるか無いかを true / false として
編集中かを判定させました。

yamaryo416 commented on src/App.jsx:
```diff
@@ -1,33 +1,100 @@
+            <TodoItem
+              todo={todo}
+              editProps={{
+                isEditing: isEditing,
+                currentTodo: currentTodo,
+                onEditTodo: (e) => setCurrentTodo({...currentTodo, text: e.target.value}),
+                cancelEdit: () => cancelEdit(),
+                updateTodos: () => updateTodos(),
```
keyとvalue名が完全一致している場合は、省略できます。
```
editProps={{
  isEditing, <= isEditingは削除する予定, TodoItem内で、そのTodoが編集中かどうかを判定する
  currentTodo,
  onEditTodo, <= こちら定義しましょう。
   cancelEdit,
  updateTodo,
```

yuyakondo333 replied:
isEditingを削除して editProps の中身を省略しました。

yuyakondo333 committed:
refactor: apply code review feedback

- Change todo count functions to constants
- Add early return pattern in addTodo
- Add key prop and remove explicit return in map
- Remove unused key prop from EditingTodo component
- Use shorthand syntax for matching key-value pairs in editProps

yuyakondo333 committed:
Removed the isEditing state variable. Use currentTodo to determine whether editing is in progress.

yamaryo416 reviewed (approved):
LGTM!
